### PR TITLE
docs: basic overview of git workflow

### LINKED ISSUE #6: git tutorials

### PR DESCRIPTION
**Pull request by exu3**
We created a markdown file that briefly covers the following: clone, pull, commit, push, pull request, and merge conflicts. 

Fixes #6

### CODE REVIEW

**Comment by awlevin** [comment]
Oops didn’t mean to mark this as ready for review! Sorry bout that, the button was just so big and right where I was scrolling on mobile 😅

**Comment by awlevin** [comment]
This looks fantastic so far!! Great work 😄

File: git-tutorials/git.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+First things first, we'll want to get a copy of our repository on our computer locally by cloning our repository. We will need the HTTPS address on the repository page which you can copy/paste from the browser. We will now use our terminal's shell commands to move to the place where we want to put our directory copy.
+
+For example, if we wanted to change to our Documents directory, we would do the following:
+`cd Documents`
+
+To clone our repository, we use the following command in our terminal:
+`git clone <https address here>`
+
+For example:
+`git clone https://github.com/ThinkFiveable/open fiveable-open`
```
Comment (zaida04) [inline git-tutorials/git.md]:
```suggestion
`git clone https://github.com/YOURUSERNAMEHERE/open.git fiveable-open`
```

Edit: I originally had this edit change it to: `git clone https://github.com/ThinkFiveable/open.git fiveable-open`, but then I wrote https://github.com/ThinkFiveable/open/pull/31#issuecomment-808788686

File: git-tutorials/git.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+Again, this will create a local copy of the repository at the directory specified earlier. You will then need to open these files in your preferred text editor to make your changes.
+ 
+### Pull
+You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 
+
+In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line.
+First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash pop` to apply the changes back onto your local copy.
```
Comment (zaida04) [inline git-tutorials/git.md]:
```suggestion
First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack.
```
https://codingkilledthecat.wordpress.com/2012/04/27/git-stash-pop-considered-harmful/.

In short, unexpected behavior can occur in the event of pull/merge conflicts, which means git stash pop won't actually remove from the stack like expected. IMO it'd be better to be consistent to begin with than to have to teach edge cases like that.

File: git-tutorials/git.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+- [Getting Started with Git/GitHub](#getting-started-with-git-github)
+    + [Clone](#clone)
+    + [Pull](#pull)
+    + [Commit](#commit)
+    + [Push](#push)
+    + [Pull Request](#pull-request)
+
+# Getting Started with Git/GitHub
+To use Git, you'll need to download it [here](https://git-scm.com/downloads).
```
Comment (zaida04) [inline git-tutorials/git.md:12]:
```suggestion
To use Git, you'll need to download it [here](https://git-scm.com/downloads).

Be sure to set your username and email before making contributions.
* `git config --global user.name "Your Name"`
* `git config --global user.email "youremail@yourdomain.com"`

File: git-tutorials/git.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+### Pull
+You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 
+
+In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line.
+First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash pop` to apply the changes back onto your local copy.
+
+### Commit
+After making your changes, you will use terminal commands to commit them. Our commit message needs to explain what we've done. This message may be useful if we want to check our change history in the future, and it can also be useful to other contributors.
+
+The syntax for a commit message is as follows `git commit -m "your commit message here"`.
```
Comment (zaida04) [inline git-tutorials/git.md:55]:
```suggestion
The syntax for a commit message is as follows `git commit -m "your commit message here"`.
For our repository, please ensure your commit messages follow [this standard](https://github.com/ThinkFiveable/open/blob/main/CONTRIBUTING.md#commiting)
```

**Review by zaida04 — CHANGES_REQUESTED** [review]

**Comment by zaida04** [comment]
~~In addition to the changes above, there was one problem that I noticed will raise an issue later.~~ I added a comment referring to this comment at: https://github.com/ThinkFiveable/open/pull/31#discussion_r602779499

The clone URL that you provide (`https://github.com/ThinkFiveable/open`) will set the main remote url to the *main* repository. This will cause an issue when someone tries to do `git push`, as git will attempt to push to **this** repository, instead of a fork of their own repository. 

## So how do we solve this??
We need a `fork` section of this guide which teaches people how to fork the repository and then we need to edit the clone section appropriately to use THEIR forks URL. It will look something like `git clone https://github.com/URNAMEHERE/open`.

**Comment by zaida04** [comment]
And (last comment I swear 😛), we should encourage people to switch to a new branch when they make new changes instead of commiting to their master branch. 

## Why?
Say you commit to the master branch of your fork. You go ahead and make a PR, but oh no! Your PR was denied and was closed. You now have a problem, **your master branch now has a bunch of commits that don't exist on the upstream repo. Your fork is now out of sync with the upstream repo.** You *could* just revert those commits, but now you have a bunch of revert commits polluting your commit history, and those will be added to the upstream repo if you make another PR. Or you could rebase and remove history (which would a require a force push which we do NOT allow on this repo). Both of these options are not very good and will only cause hassle to the maintainers and to the contributor themselves.

Sounds like a lot right? It's better to prevent all of this from happening in the beginning by just **making changes to a new branch every time**. So say I want to add a bunch of badges to the readme, i'd make a branch `chore/readme-branch` or `chore/patch-1`.

## So how would we solve this?
Add a `checkout` section to this guide that shows people how to switch to a new branch, create a new branch, and whatnot. 
*Also a nice little trick is `git switch branch-name` and `git switch -c branch-name`*. Say you made a bunch of commits, but you forgot to switch branches. Instead of stashing and applying or discarding and redoing, you can just do the first or second command. The differences between the two are the first one will switch only to an existing branch, while the second one will create the branch if it doesn't exist. In the end, these both will **move** your already done commits to this new branch. Ref: https://git-scm.com/docs/git-switch

File: git-tutorials/git.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+First, we will need our own copy of the repository to make changes on. You will need to create a fork by clicking on the `Fork` button in the top right.
+
+### Clone
+Next, we'll want to get a copy of our repository on our computer locally by cloning our repository. We will need the HTTPS address on the repository page which you can copy/paste from the browser. We will now use our terminal's shell commands to move to the place where we want to put our directory copy.
+
+For example, if we wanted to change to our Documents directory, we would do the following:
+`cd Documents`
+
+To clone our repository, we use the following command in our terminal:
+`git clone <https address here>`
```
Comment (zaida04) [inline git-tutorials/git.md]:
```suggestion
`git clone <https address here> [cloned-directory-name]`
```

File: git-tutorials/git.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+* `git checkout <branchname>` to swtich to your branch
+
+If you make a bunch of commits and forget to switch branches, you can use `git switch <branchname>` or `git switch -c <branchname>` which will switch to another branch and move your already-done commits to this new branch. Appending `-c` will create a new branch while `git switch <branchname>` switches to an existing branch.
+ 
+### Pull
+You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 
+
+In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. This method is not reccommened as unexpected behavior may occur.
```
Comment (zaida04) [inline git-tutorials/git.md]:
```suggestion
In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. 
```
💀 I think there was a mixup, `git stash pop` is what is not recommended and what'll cause unexpected behavior. `git stash apply` & `git stash drop` are the alternative.

**Review by zaida04 — CHANGES_REQUESTED** [review]
Other than these two changes, this PR looks good to go! :D

**Review by zaida04 — DISMISSED** [review]

**Review by sarthaktexas — DISMISSED** [review]

File: git-tutorials/git.md [hunk 7]
```diff
@@ -70,6 +70,17 @@ If you click that button, you will be asked to give your pull reqest a name and
...
 
+As your PR waits to be merged, other peoples' PRs might be merged before yours, meaning you have to "update your branch" with the master branch. You can do it with either this button 
+
+<img align="center" src="https://cdn.discordapp.com/attachments/768856355255484469/825476408767545392/chrome_2e0DN4u2iQ.png" alt="Update branch button"/>
+
+or (the more preferred method as the above button won't always be avaliable), you can do these steps:
+- Add our repository as a remote URL (`git remote add upstream https://github.com/ThinkFiveable/open.git`)
+- Make sure you are on the branch you want to merge master into (`git checkout PR_BRANCH_NAME`)
+- Make sure you have the latest upstream changes (`git fetch upstream`)
+- Merge the upstream master branch into your PR branch (`git merge upstream/master`). At this stage you might run into what are called **MERGE CONFLICTS. You must resolve these before pushing to your fork.** See: [Merge Conflicts](#merge-conflicts)
```
Comment (exu3) [inline git-tutorials/git.md]: Should it be `git merge upstream/main`?

Comment (zaida04) [reply git-tutorials/git.md]: Yeah totally forgot we use main

**Review by zaida04 — APPROVED** [review]

File: git-tutorials/git.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+* `git branch <branchname>` to create a new branch
+* `git checkout <branchname>` to swtich to your branch
+
+If you make a bunch of commits and forget to switch branches, you can use `git switch <branchname>` or `git switch -c <branchname>` which will switch to another branch and move your already-done commits to this new branch. Appending `-c` will create a new branch while `git switch <branchname>` switches to an existing branch.
+ 
+### Pull
+You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 
+
+In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack.
```
Comment (sarthaktexas) [inline git-tutorials/git.md:50]: why don't you just do git stash pop??

Comment (zaida04) [reply git-tutorials/git.md:50]: https://github.com/ThinkFiveable/open/pull/31#discussion_r602780056

**Review by sarthaktexas — APPROVED** [review]

**Pull request merged**